connmgr/connmanager_test.go

```diff
@@ -0,0 +1,373 @@
+	gotID = gotConnReq.ID()
+	if gotID != wantID {
+		t.Fatalf("retry: %v - want ID %v, got ID %v", cr.Addr, wantID, gotID)
+	}
+	gotState = cr.State()
+	wantState = connmgr.ConnDisconnected
+	if gotState != wantState {
+		t.Fatalf("retry: %v - want state %v, got state %v", cr.Addr, wantState, gotState)
+	}
+
+	gotConnReq = <-connected
+	wantID = cr.ID()
+	gotID = gotConnReq.ID()
+	if gotID != wantID {
+		t.Fatalf("retry: %v - want ID %v, got ID %v", cr.Addr, wantID, gotID)
+	}
+	gotState = cr.State()
+	wantState = connmgr.ConnEstablished
+	if gotState != wantState {
+		t.Fatalf("retry: %v - want state %v, got state %v", cr.Addr, wantState, gotState)
+	}
+
+	cmgr.Remove(cr.ID())
+	gotConnReq = <-disconnected
+	wantID = cr.ID()
+	gotID = gotConnReq.ID()
+	if gotID != wantID {
+		t.Fatalf("retry: %v - want ID %v, got ID %v", cr.Addr, wantID, gotID)
+	}
+	gotState = cr.State()
+	wantState = connmgr.ConnDisconnected
+	if gotState != wantState {
+		t.Fatalf("retry: %v - want state %v, got state %v", cr.Addr, wantState, gotState)
+	}
+	cmgr.Stop()
+}
+
+// TestMaxRetryDuration tests the maximum retry duration.
+//
+// We have a timed dialer which initially returns err but after RetryDuration
+// hits maxRetryDuration returns a mock conn.
+func TestMaxRetryDuration(t *testing.T) {
+	networkUp := make(chan struct{})
+	time.AfterFunc(5*time.Millisecond, func() {
+		close(networkUp)
+	})
+	timedDialer := func(network, address string) (net.Conn, error) {
+		select {
+		case <-networkUp:
+			return mockDialer(network, address)
+		default:
+			return nil, errors.New("network down")
+		}
+	}
+
+	connected := make(chan *connmgr.ConnReq)
+	cmgr, err := connmgr.New(&connmgr.Config{
+		RetryDuration: time.Millisecond,
+		MaxOutbound:   1,
+		Dial:          timedDialer,
+		OnConnection: func(c *connmgr.ConnReq, conn net.Conn) {
+			connected <- c
+		},
+	})
+	if err != nil {
+		t.Fatalf("connmgr.New error: %v", err)
+	}
+
+	cr := &connmgr.ConnReq{Addr: "127.0.0.1:18555", Permanent: true}
+	go cmgr.Connect(cr)
+	cmgr.Start()
+	// retry in 1ms
+	// retry in 2ms - max retry duration reached
+	// retry in 2ms - timedDialer returns mockDial
+	<-connected
+}
+
+// TestNetworkFailure tests that the connection manager handles a network
+// failure graciously.
+func TestNetworkFailure(t *testing.T) {
+	var dials uint32
+	errDialer := func(network, address string) (net.Conn, error) {
+		atomic.AddUint32(&dials, 1)
+		return nil, errors.New("network down")
+	}
+	cmgr, err := connmgr.New(&connmgr.Config{
+		MaxOutbound:   5,
+		RetryDuration: 5 * time.Millisecond,
+		Dial:          errDialer,
+		GetNewAddress: func() string { return "127.0.0.1:18555" },
+		OnConnection: func(c *connmgr.ConnReq, conn net.Conn) {
+			t.Fatalf("network failure: got unexpected connection - %v", c.Addr)
+		},
+	})
+	if err != nil {
+		t.Fatalf("connmgr.New error: %v", err)
+	}
+	cmgr.Start()
+	time.AfterFunc(10*time.Millisecond, cmgr.Stop)
+	cmgr.Wait()
+	if atomic.LoadUint32(&dials) > 25 {
```

Comment:
25 seems a bit high here.  There should only be 5 simultaneous attempts with a 5 millisecond retry.  Thus there will be 5 attempts, then 5 milliseconds later another 5 attempts, and then 5 more attempts after 10 milliseconds.  The last one is unlikely to be hit since it's stopped after 10 milliseconds, but to be safe I think 15 would cover it.